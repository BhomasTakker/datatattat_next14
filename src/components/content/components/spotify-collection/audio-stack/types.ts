import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
