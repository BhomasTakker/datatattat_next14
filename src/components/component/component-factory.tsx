import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/component/component-map";
import { PageComponent } from "@/types/page";

export const ComponentFactory = ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>Component not found</div>;
	}

	return <Component component={component} />;
};
