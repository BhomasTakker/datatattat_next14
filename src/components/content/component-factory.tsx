import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/content/component-map";
import { PageComponent } from "@/types/page";

export const ComponentFactory = ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>Component not found</div>;
	}

	return <Component component={component} />;
};
