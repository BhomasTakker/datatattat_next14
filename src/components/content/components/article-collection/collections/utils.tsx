import { StyleSheet } from "@/types/css";
import { Article } from "../article/article";
import { ArticleRenderProps } from "./types";
import { getClientMeta } from "@/actions/client/get-meta";

// Somwhere better
export const articleRenderer =
	(styles: StyleSheet) => (item: ArticleRenderProps) =>
		<Article article={item} styles={styles} />;

export const articleTemplate = (styles: StyleSheet) => (
	<div className={styles.template} />
);

export const articleMetaLoader = (item: ArticleRenderProps) => async () => {
	const loadData = item?.loadData;
	if (loadData) {
		const data = await getClientMeta(item);
		return data;
	}
	return item;
};
