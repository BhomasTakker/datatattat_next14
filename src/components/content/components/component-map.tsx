import { ArticleCollection } from "./article-collection/article-collection";
import { ContentOembed } from "./content-oembed/content-oembed";
import { DisplayPlayer } from "./display-player/display-player";
import { MediaPlayer } from "./media-player/media-player";

export type ComponentType =
	| typeof ArticleCollection
	| typeof ContentOembed
	| typeof DisplayPlayer
	| typeof MediaPlayer;

export enum ComponentsOptions {
	ArticleCollection = "ArticleCollection",
	ContentOembed = "ContentOembed",
	DisplayPlayer = "DisplayPlayer",
	MediaPlayer = "MediaPlayer",
}

export const ComponentsMap = new Map<ComponentsOptions, ComponentType>([
	[ComponentsOptions.ArticleCollection, ArticleCollection],
	[ComponentsOptions.ContentOembed, ContentOembed],
	[ComponentsOptions.DisplayPlayer, DisplayPlayer],
	[ComponentsOptions.MediaPlayer, MediaPlayer],
]);
