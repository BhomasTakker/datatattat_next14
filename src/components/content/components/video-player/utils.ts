import Player from "video.js/dist/types/player";
import { PlayerOptions } from "./types";

// Probably do something better than this
// Just to avoid dupe code
export const applyPlayerOptions = (player: Player, options: PlayerOptions) => {
	const {
		autoplay,
		aspectRatio,
		controls,
		fill,
		fluid,
		responsive,
		poster,
		audioPosterMode,
		playbackRates,
		width,
		height,
		preload,
		techOrder,
		sources,
	} = options;

	player.fill(fill);
	player.controls(controls);
	player.aspectRatio(aspectRatio);
	player.autoplay(autoplay);
	player.poster(poster);

	// if (options.autoplay) {
	// 	player.play();
	// }
};
