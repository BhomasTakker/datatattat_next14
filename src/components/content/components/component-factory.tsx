import { getData } from "@/actions/data/get-data";
import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/content/components/component-map";
import { PageComponent } from "@/types/page";
import { UnknownObject } from "@/types/utils";

export const ComponentFactory = async ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>{`Component not found:- ${componentType}`}</div>;
	}

	const { _with: queryObject } = component;

	const data = (await getData(queryObject)) as UnknownObject;

	// all components take in a component object
	// AND and a data object
	// data object contains data and any data functions
	// data functions can be used to refecth or manipulate queries
	return <Component component={component} dataObject={{ data }} />;
};
