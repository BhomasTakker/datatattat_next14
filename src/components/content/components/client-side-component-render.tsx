"use client";

import { PageComponent } from "@/types/page";
import { UnknownObject } from "@/types/utils";
import { useEffect, useState } from "react";
import { ComponentType } from "./component-map";
import { With } from "@/types/component";

type ClientSideComponentProps = {
	Component: ComponentType;
	component: PageComponent;
	getData: (queryObject: With) => Promise<UnknownObject>;
};

export const ClientSideComponent = ({
	getData,
	Component,
	component,
}: ClientSideComponentProps) => {
	const { _with: componentQueryObject, componentProps } = component;
	const { _with: componentPropsQueryObject } = componentProps || {};
	const [componentData, setComponentData] = useState<UnknownObject | null>(
		null
	);
	const [isError, setIsError] = useState<boolean>(false);

	useEffect(() => {
		const getComponentData = async () => {
			try {
				const data = await getData(
					componentQueryObject || componentPropsQueryObject
				);
				setComponentData(data);
			} catch {
				setIsError(true);
			}
		};
		getComponentData();
	}, []);

	if (isError) {
		return <div>Error loading data</div>;
	}

	if (!componentData) {
		return <div>Loading...</div>;
	}

	return (
		<Component component={component} dataObject={{ data: componentData }} />
	);
};
