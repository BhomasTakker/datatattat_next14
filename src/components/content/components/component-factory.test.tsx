import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

import { ComponentFactory } from "../components/component-factory";

import { ComponentsOptions } from "../components/component-map";
import { PageComponent } from "@/types/page";
import { getCurrentRoute } from "../../../utils/route";

// Would need to mock a couple of options to test the default component

const MOCK = {
	component: {
		componentType: ComponentsOptions.ArticleCollection,
	},
};

jest.mock("../../../actions/data/get-data", () => {
	return {
		getData: jest.fn().mockResolvedValue({}),
	};
});

jest.mock("../../../utils/route", () => {
	return {
		getCurrentRoute: jest.fn().mockResolvedValue("/"),
	};
});

jest.mock("./article-collection/article-collection", () => {
	return {
		__esModule: true,
		ArticleCollection: () => {
			return <div data-testid="article-collection">Article Collection</div>;
		},
	};
});

describe("ComponentFactory", () => {
	it("should render the Article Collection component", async () => {
		render(
			await ComponentFactory({ component: MOCK.component as PageComponent })
		);
		const component = screen.getByText("Article Collection");
		expect(component).toBeInTheDocument();
	});

	it("should render a default component if the component type is not found", async () => {
		// @ts-expect-error - We are testing the default component
		MOCK.component.componentType = "Unknown";
		render(
			await ComponentFactory({ component: MOCK.component as PageComponent })
		);
		const component = screen.getByText(/Component not found/i);
		expect(component).toBeInTheDocument();
	});

	it("renders ComponentFactory unchanged", async () => {
		const { container } = render(
			await ComponentFactory({ component: MOCK.component as PageComponent })
		);
		expect(container).toMatchSnapshot();
	});
});
