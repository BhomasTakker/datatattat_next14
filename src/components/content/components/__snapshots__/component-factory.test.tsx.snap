// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ComponentFactory renders ComponentFactory unchanged 1`] = `
<div>
  <div>
    Component not found:- Unknown
  </div>
</div>
`;
